### PR TITLE
Better default expiration strategy

### PR DESCRIPTION
**Pull request by fuchsnj**
The default expiration strategy was to never expire things. If you didn't override this, that means that if the sqs consumer is slow then you could end up processing messages that are either already expired, or so close to expiring that it will expire before you finish processing. This isn't a great default, so it was changed to a strategy that expires it if the age of the message is over 50% of the visibility timeout of the queue before it starts processing.

This also changes the default MessageAcknowledger from expiring after the visiblity timeout, to never expiring. This can cause a subtle bug where your consumer ends up receiving more messages than expected if a message expires while it's being processed. You can still turn this feature on in the SqsConsumerBuilder, but the default is now off.

### CODE REVIEW

File: src/main/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpiration.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+public class VisibilityTimeoutPercentageExpiration implements ExpirationStrategy {
+
+    private final double percent;
+
+    public VisibilityTimeoutPercentageExpiration(double percent) {
+        this.percent = percent;
+    }
+
+    @Override
+    public boolean isExpired(SqsMessage<?> sqsMessage, Duration visibilityTimeout) {
```
Comment (derelbenkoenig) [inline src/main/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpiration.java]: Couldn't the visibility timeout just be in the constructor of this implementation? Do we expect it to change during runtime?

Comment (fuchsnj) [reply src/main/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpiration.java]: This strategy could be used for multiple queues (with different visibility timeouts)

Comment (fuchsnj) [reply src/main/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpiration.java]: And in the future it would be good to detect changes in visibility timeout, and these could be applied to the expiration strategy automatically

Comment (derelbenkoenig) [reply src/main/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpiration.java]: I would assume that if you were using it with multiple queues, you would just make a separate instance of the `ExpirationStrategy` for each one. If you could detect changes in it then it makes more sense, but it might seem better to have something able to set it on the class instead of passing it to each call, if that matters enough

**Comment by tpensyl** [comment]
As you noted in option B), once we pass the visibility timeout from SQS, we can no longer tell it to delete the message from SQS, so we can't avoid the retry, so option B only helps if message is processed after our internal expire but before the SQS visibility timeout. The extra complexity doesn't seem worth handling the edge case only in this limited scenario.
We should still evaluate our deadletter policies to make sure they have enough retries to minimize chance of bad deadlettering.

File: src/test/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpirationTest.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    @Test
+    public void testIsExpiredFalse() {
+        SqsMessage<String> message = SqsMessage.<String>builder()
+                .body(MESSAGE_BODY)
+                .id(MESSAGE_ID)
+                .receivedTime(Instant.now())
+                .receiptHandle(RECEIPT_HANDLE)
+                .build();
+
+        assertThat(strategy.isExpired(message, VISIBILITY_TIMEOUT)).isFalse();
```
Comment (derelbenkoenig) [inline src/test/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpirationTest.java]: This would fail if the test took more than 3 minutes.
You should have the strategy take a `Clock` as one of its constructor params, and in the actual application use the regular `systemUTC` clock, but for this test use a mocked out one

File: src/main/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpiration.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package com.bandwidth.sqs.consumer.strategy.expiration;
+
+import com.bandwidth.sqs.queue.SqsMessage;
+
+import java.time.Duration;
+
+/**
+ * Considers a message expired after it's age is a certain percentage of the visibility timeout
```
Comment (derelbenkoenig) [inline src/main/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpiration.java]: "it's" -> "its"

Comment (tpensyl) [reply src/main/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpiration.java]: Bikeshedding Benson

Comment (derelbenkoenig) [reply src/main/java/com/bandwidth/sqs/consumer/strategy/expiration/VisibilityTimeoutPercentageExpiration.java]: I don't know if that fits the definition of bikeshedding

File: src/main/java/com/bandwidth/sqs/consumer/SqsConsumerBuilder.java [hunk 4]
```diff
@@ -20,7 +21,7 @@
     int bufferSize = DEFAULT_BUFFER_SIZE;
     int priority = DEFAULT_PRIORITY;
     BackoffStrategy backoffStrategy = new NullBackoffStrategy();
-    ExpirationStrategy expirationStrategy = new NeverExpiresStrategy();
+    ExpirationStrategy expirationStrategy = new VisibilityTimeoutPercentageExpiration(0.5);
```
Comment (derelbenkoenig) [inline src/main/java/com/bandwidth/sqs/consumer/SqsConsumerBuilder.java]: `DEFAULT_VISIBILITY_TIMEOUT_RATIO`

File: src/main/java/com/bandwidth/sqs/consumer/acknowledger/MessageAcknowledger.java [hunk 5]
```diff
@@ -119,7 +119,10 @@ public void transfer(T newMessage, SqsQueue<T> newQueue){
 
     private Completable doTransfer(T newMessage, SqsQueue<T> newQueue, Optional<Duration> delay) {
         return newQueue.publishMessage(newMessage, delay)
-                .flatMapCompletable((msgId) -> sqsQueue.deleteMessage(receiptId));
+                .flatMapCompletable((msgId) -> {
+                    Completable completable = sqsQueue.deleteMessage(receiptId);
```
Comment (derelbenkoenig) [inline src/main/java/com/bandwidth/sqs/consumer/acknowledger/MessageAcknowledger.java]: what

File: src/test/java/com/bandwidth/sqs/consumer/SqsConsumerTest.java [hunk 6]
```diff
@@ -205,11 +206,53 @@ public void testProcessNextMessage() {
...
+                .from(SQS_MESSAGE)
+                .receivedTime(Instant.now().minus(ATTRIBUTES.getVisibilityTimeout()).plus(Duration.ofSeconds(6)))
+                .build());
+
+        consumer.setMessageBuffer(messageBuffer);
+        consumer.processNextMessage(consumer.getNextMessage());
+
+        verify(consumerHandlerMock, never()).handleMessage(eq(SQS_MESSAGE), any());
+        verify(sqsQueueMock).publishMessage(any(), any()); //make sure a new message is published
+        verify(sqsQueueMock).deleteMessage(anyString());// ...and the old one is deleted
```
Comment (derelbenkoenig) [inline src/test/java/com/bandwidth/sqs/consumer/SqsConsumerTest.java]: make sure it deletes the correct one instead of `anyString()`?
